src/server/pkg/storage/fileset/storage.go

```diff
@@ -328,30 +274,60 @@ func (s *Storage) GC(ctx context.Context) error {
 	return gc.Run(ctx)
 }
 
-func (s *Storage) levelSize(i int) int64 {
-	return s.levelZeroSize * int64(math.Pow(float64(s.levelSizeBase), float64(i)))
-}
-
-const subFileSetFmt = "%020d"
-const levelFmt = "level_" + subFileSetFmt
-
-// SubFileSetStr returns the string representation of a subfileset.
-func SubFileSetStr(subFileSet int64) string {
-	return fmt.Sprintf(subFileSetFmt, subFileSet)
+func (s *Storage) newPrimitive(ctx context.Context, prim *Primitive, ttl time.Duration) (*ID, error) {
+	id := newID()
+	md := &Metadata{
+		Value: &Metadata_Primitive{
+			Primitive: prim,
+		},
+	}
+	var pointsTo []string
+	for _, chunkID := range prim.PointsTo() {
+		pointsTo = append(pointsTo, chunk.ObjectID(chunkID))
+	}
+	if err := s.tracker.CreateObject(ctx, filesetObjectID(id), pointsTo, ttl); err != nil {
+		return nil, err
+	}
+	if err := s.store.Set(ctx, id, md); err != nil {
+		return nil, err
+	}
+	return &id, nil
 }
 
-func levelName(i int) string {
-	return fmt.Sprintf(levelFmt, i)
+func (s *Storage) newComposite(ctx context.Context, comp *Composite, ttl time.Duration) (*ID, error) {
+	id := newID()
+	md := &Metadata{
+		Value: &Metadata_Composite{
+			Composite: comp,
+		},
+	}
+	var pointsTo []string
+	for _, id := range comp.Layers {
+		pointsTo = append(pointsTo, filesetObjectID(id))
+	}
+	if err := s.tracker.CreateObject(ctx, filesetObjectID(id), pointsTo, ttl); err != nil {
+		return nil, err
+	}
+	if err := s.store.Set(ctx, id, md); err != nil {
+		return nil, err
+	}
+	return &id, nil
```

Comment:
Why do we need to return a pointer here? If we can just return the ID, then that seems preferable.

Reply:
I'm thinking of making the ID a UUID. Either way I want it to be opaque to the caller. I would rather code outside the fileset package return nil, than have to know IDs are strings at the moment.